teps/0091-verified-task-bundles.md

```diff
@@ -0,0 +1,536 @@
+attestation claims that are made.
+
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Provide solid building blocks to begin to establish formal chains of trust
+  within the build pipeline itself.
+- Leveraging the above, provide mechanisms to establish a verifiable corpus
+  of community-provided Task Bundles.
+- Create an optional Tekton configuration for Task Bundle verification based
+  on Sigstore Cosign (https://github.com/sigstore/cosign).
+- Create an optional Tekton Chains configuration to skip attestation creation
+  based on information from Task Bundles that could not be verified.
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Specification of a verification mechansim for arbitrary Tasks.
+- Specification of a verification mechanism for the images within a Task
+  Bundle.
+
+
+### Use Cases (optional)
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+**Kaniko builds.** A Kaniko task exists in the Tekton Catalog. This task emits
+output that allows Chains to make attestations. The task uses the
+`gcr.io/kaniko-project/executor` to perform the build and push the resulting
+image. But how can I be sure this Task is not compromised and is actually doing
+what I expect? At a minimum, I would like to be able to verify it, for example,
+using Sigstore's Cosign, to know that it was produced by the expected
+third-party.
+
+**Buildpack builds.** Similar to Kaniko, a Buildpacks task exists in the Tekton
+Catalog with output that indicates the image that was built. The motivation is
+the same here: how can I trust the Task without some sort of verification?
+
+**Really, any third-party task.** If we're going to build an ecosystem of tasks
+in the Tekton Catalog, we need to think about how to ensure these tasks are
+safe. Task verification can provide building blocks.
+
+**First-party tasks.** In the spirit of zero-trust, I'd like to sign my own
+Tasks and verify them on each use. This would be a best practice for
+enterprises, for example.
+
+
+## Requirements
+
+<!--
+Describe constraints on the solution that must be met. Examples might include
+performance characteristics that must be met, specific edge cases that must
+be handled, or user scenarios that will be affected and must be accomodated.
+-->
+
+- A Pipeline Author can indicate that a Task Bundle must be verified in order
+  for Chains to create a provenance attestation.
+- A Pipeline Author can configure a build to fail if a Task Bundle cannot be
+  verified.
+
+
```

Comment:
> Are there other types of annotations in Tekton that are subject to abuse if these annotations can be edited like this?

i think you could make things not work correctly, e.g. by changing labels or even changing owner references, but im not sure what the abuse vector would be (general disruption and mischief?)

I believe tekton chains signs the annotations it adds ( @priyawadhwa correct me if i'm wrong ) - maybe the same approach will work for pipelines adding annotations we need to trust?

Reply:
@bobcatfish yah so the annotations could be edited to say pretty much anything. If you want to prevent that then this is where spire could come in -- with spire, the pipelines controller could add some annotation and also a signature over that annotation, which chains could then verify. you can configure spire to only sign requests that come from the tekton controller, so something else running the cluster wouldn't be able to do it.

to follow the current chains format for signing you would have the user specify a path to a private key (this could be in a k8s secret or in kms). you'd just have to make sure that nothing else in the cluster could sign with it!